### PR TITLE
LICM: add an optimization to move multiple loads and stores from/to the same memory location out of a loop.

### PR DESCRIPTION
**Pull request by eeckstein**
This is a combination of load hoisting and store sinking, e.g.
```
  preheader:
    br header_block
  header_block:
    %x = load %not_aliased_addr
    // use %x and define %y
    store %y to %not_aliased_addr
    ...
  exit_block:
```
is transformed to:
```
  preheader:
    %x = load %not_aliased_addr
    br header_block
  header_block:
    // use %x and define %y
    ...
  exit_block:
    store %y to %not_aliased_addr
```

This optimization is important to optimize inout arguments, especially with COW support in SIL: it relies on values being in SSA values rather than in memory.

This PR is for test and review for now. If everything goes well I'll merge it next week.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci test

File: lib/SILOptimizer/LoopTransforms/LICM.cpp [hunk 1]
```diff
@@ -737,14 +819,185 @@ bool LoopTreeOptimization::optimizeLoop(
...
+      // If there are multiple stores in a block, only the last one counts.
+      loc = SI->getLoc();
+      ssaUpdater.AddAvailableValue(SI->getParent(), SI->getSrc());
+    }
+  }
+
+  // Remove all stores and replace the loads with the current value.
+  SILBasicBlock *currentBlock = nullptr;
+  SILValue currentVal;
+  for (SILInstruction *I : LoadsAndStores) {
```
Comment (atrick) [inline lib/SILOptimizer/LoopTransforms/LICM.cpp]: This has an assumption that LoadsAndStores are in program order but there's no contract. It isn't specified at the declaration or when the LoadsAndStores list is created.

**Review by atrick — COMMENTED** [review]
This is really good. I have a few comments below though.

And I just realized this PR isn't the code that I reviewed, so I'll add another comment here
https://github.com/apple/swift/pull/27990

**Pull request merged**